src/Microsoft.ML/Runtime/Internal/Tools/CSharpApiGenerator.cs

```diff
@@ -495,16 +495,56 @@ private void GenerateFooter(IndentingTextWriter writer)
             writer.WriteLine();
         }
 
-        static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, string fullTypeName, string currentNamespace)
+        static string GetSymbolFromType(Dictionary<string, string> typesSymbolTable, Type type, string currentNamespace)
         {
+            var fullTypeName = type.FullName;
             var names = typesSymbolTable.Select(kvp => kvp.Value);
             char dim = fullTypeName.Contains('+') ? '+' : '.';
             string name = currentNamespace != "" ? currentNamespace + '.' : "";
 
+            Type[] genericTypes = null;
+            if (type.IsGenericType)
+                genericTypes = type.GetGenericArguments();
+
             if (fullTypeName.Contains('+'))
-                name += fullTypeName.Substring(0, fullTypeName.LastIndexOf('+')).Substring(fullTypeName.LastIndexOf('.') + 1);
+            {
+                int bracketIndex = fullTypeName.IndexOf('[');
+                if (bracketIndex > 0)
+                {
+                    Contracts.AssertValue(genericTypes);
+                    fullTypeName = fullTypeName.Substring(0, bracketIndex);
+                }
```

Comment:
This is some fairly elaborate string manipulation going on, the utility of which will be opaque to anyone not intimately familiar with how the .NET class naming system works, and that also does not understand the specific issues raised in issue #101 that requires its introduction.

Reply:
I have a feeling this is more complicated. For comparison, there's a similar case (maybe not exactly, but anyway): https://github.com/dotnet/orleans/blob/master/src/AdoNet/Orleans.Persistence.AdoNet/Storage/Provider/AdoNetGrainStorage.cs#L506. More laborius, but might be usable to extract this to a separate class/function and test a bit more. This is something that probably should be done in Orleans too, since there are similar needs. :)

Also perhaps interesting in case there's a possibility to refactor out commonalities and have a joint library: https://github.com/dotnet/orleans/blob/e78c7fa4ae302808ed30d71ebc14f1ef544a923e/src/Orleans.Core/AssemblyLoader/AssemblyLoader.cs.

Reply:
Hi Veikko, thanks for your suggestion and for the reference. This method definitely doesn't handle every possible case for a type name (for example, it does not handle the case where the nested class is also generic). Since it is a private method, I think that would be enough for handle the cases we need it to handle. I did add some comments though, to make it easier to handle new cases if the need arises.

Reply:
I might be tempted to insert a link though to the MSDN docs for this at the GetType, since we're attempting to read a (subset) of the language. Plus maybe document the limitations of this method.

---
In reply to: [187414843](https://github.com/dotnet/machinelearning/pull/102#discussion_r187414843) [](ancestors = 187414843)